### PR TITLE
#1 - Nuspec parsing & saving

### PR DESCRIPTION
**Pull request by olegmoz**
This is part of issue: https://github.com/artipie/nuget-adapter/issues/1
Adding package to repository includes saving .nuspec manifest extracted from package to package path.

### CODE REVIEW

**Comment by olegmoz** [comment]
@g4s8 please assign reviewer for this PR

**Comment by g4s8** [comment]
@0crat in

**Comment by 0crat** [comment]
> @0crat in ([here](https://github.com/artipie/nuget-adapter/issues/10#issuecomment-581915603))

@g4s8 Job [#10](https://github.com/artipie/nuget-adapter/issues/10) is already in scope

**Comment by g4s8** [comment]
@0crat assign @carlosmiranda

**Comment by 0crat** [comment]
> @0crat assign @carlosmiranda  ([here](https://github.com/artipie/nuget-adapter/issues/10#issuecomment-581945887))

@g4s8 There is an unrecoverable failure on my side. Please, submit it [here](https://github.com/zerocracy/farm/issues):

```
PID: 11@377a78bde909, thread: PQ-CT2E6TK9B
com.zerocracy.claims.ClaimIn[214] java.lang.IllegalArgumentException: Parameter "age" not found in "Order was canceled" among: [comment, job, issue, repo, cause, _parent_sig, flow, login]
```

[0.51.0](https://github.com/zerocracy/farm): CID: [5b112376-139e-4f05-ad76-459407c66e39](https://www.0crat.com/footprint/CT2E6TK9B/5b112376-139e-4f05-ad76-459407c66e39), Type: "Order was canceled"

**Comment by 0crat** [comment]
> @0crat assign @carlosmiranda  ([here](https://github.com/artipie/nuget-adapter/issues/10#issuecomment-581945887))

@g4s8 This pull request [#10](https://github.com/artipie/nuget-adapter/issues/10) is assigned to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/2dc4f4a0-5dce-4807-80df-0f1960ae12f3); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by 0crat** [comment]
This pull request [#10](https://github.com/artipie/nuget-adapter/issues/10) is assigned to @iakunin[/z](https://www.0crat.com/u/iakunin), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/91b0081a-4e82-401b-825b-4952552497bb); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by 0crat** [comment]
Manual assignment of issues is discouraged, see [§19](http://www.zerocracy.com/policy.html#19): -5 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Comment by 0crat** [comment]
@0crat[/z](https://www.0crat.com/u/0crat) I see that you unassigned this issue; the order is still assigned to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda) though; to cancel the order use `refuse`, as in [§6](http://www.zerocracy.com/policy.html#6)

**Comment by 0crat** [comment]
Job [#10](https://github.com/artipie/nuget-adapter/issues/10) is already in the agenda of @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda)

**Comment by yegor256** [comment]
@iakunin it's yours, I believe

**Comment by yegor256** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/artipie/nuget-adapter/issues/10#issuecomment-582261483))

@yegor256 This is what I know about this job in [CT2E6TK9B](https://www.0crat.com/p/CT2E6TK9B), as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#10](https://github.com/artipie/nuget-adapter/issues/10) is [in scope](http://www.zerocracy.com/policy.html#14) for 21hr
  * The role is `REV`
  * The job is assigned to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda) for [16hr](http://www.zerocracy.com/policy.html#8)
  * There is a monetary reward attached
  * @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda) will get 25% of their hourly rate for this job
  * The job doesn't have any [impediments](http://www.zerocracy.com/policy.html#9)
  * The [budget](http://www.zerocracy.com/policy.html#4) is 15 minutes/points
  * These users are banned and won't be assigned:
    * @olegmoz[/z](https://www.0crat.com/u/olegmoz): This user reported the ticket
    * @iakunin[/z](https://www.0crat.com/u/iakunin): User was resigned from the ticket
  * Job [footprint](https://www.0crat.com/footprint/CT2E6TK9B?q=%7Bjob%3A%27gh%3Aartipie%2Fnuget-adapter%2310%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

File: src/test/java/com/artpie/nuget/NewtonJsonPackage.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import com.google.common.io.ByteStreams;
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * Newton.Json package resources.
+ *
+ * @since 0.1
+ */
+final class NewtonJsonPackage {
```
Comment (carlosmiranda) [inline src/test/java/com/artpie/nuget/NewtonJsonPackage.java]: @olegmoz Let's avoid utility classes. See https://www.yegor256.com/2014/05/05/oop-alternative-to-utility-classes.html

Or let's just move this code into the `NuspecTest` class itself, we're only using it there anyway.

Comment (olegmoz) [reply src/test/java/com/artpie/nuget/NewtonJsonPackage.java]: Fixed, thank you

**Comment by olegmoz** [comment]
@yegor256 please review this PR

File: src/main/java/com/artpie/nuget/Nuspec.java [hunk 2]
```diff
@@ -48,6 +74,50 @@ public PackageIdentity identity() {
...
+     * Exception is thrown if zero or more then 1 values found
+     *
+     * @param xml XML document to read from.
+     * @param xpath XPath expression to select data from the XML.
+     * @return Value found by XPath
+     */
+    private static String single(final XML xml, final String xpath) {
+        final List<String> values = xml.xpath(xpath);
+        if (values.isEmpty()) {
+            final String message = String.format("No values found in path: '%s'", xpath);
```
Comment (yegor256) [inline src/main/java/com/artpie/nuget/Nuspec.java]: @olegmoz why making it a variable? Can't you just do:

```
throw new IllegalArgumentException(
  String.format("No values found in path: '%s'", xpath)
);
```

File: src/main/java/com/artpie/nuget/Nuspec.java [hunk 3]
```diff
@@ -48,6 +74,50 @@ public PackageIdentity identity() {
...
+        }
+        storage.save(this.identity().nuspecKey(), bytes);
+    }
+
+    /**
+     * Parse binary content as XML document.
+     *
+     * @return Content as XML document.
+     */
+    private XML xml() {
```
Comment (yegor256) [inline src/main/java/com/artpie/nuget/Nuspec.java]: @olegmoz you can use [jcabi-xml](https://xml.jcabi.com/), which does exactly the same, but in a more convenient way

File: src/main/java/com/artpie/nuget/PackageIdentity.java [hunk 4]
```diff
@@ -67,6 +67,16 @@ public Key hashKey() {
         );
     }
 
+    /**
+     * Get key for .nuspec file.
+     *
+     * @return Key to .nuspec file.
+     */
+    public Key nuspecKey() {
+        final String name = String.format("%s.nuspec", this.idLowerCase());
```
Comment (yegor256) [inline src/main/java/com/artpie/nuget/PackageIdentity.java]: @olegmoz you can get rid of this variable, I believe. Just use the value where it's needed. Check this out: https://www.yegor256.com/2015/09/01/redundant-variables-are-evil.html

File: src/test/java/com/artpie/nuget/NewtonJsonResource.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    NewtonJsonResource(final String name) {
+        this.name = name;
+    }
+
+    /**
+     * Reads binary data.
+     *
+     * @return Binary data.
+     */
+    public byte[] bytes() {
```
Comment (yegor256) [inline src/test/java/com/artpie/nuget/NewtonJsonResource.java]: @olegmoz you can use `ResourceOf` from Cactoos

**Comment by yegor256** [comment]
@olegmoz see my comments above

File: src/main/java/com/artpie/nuget/Nuspec.java (hunk 3 above)
Comment (olegmoz) [reply src/main/java/com/artpie/nuget/Nuspec.java]: Actually I tried it, but XMLDocument crashes with the following error in DomParser class: 
```
Doesn't look like XML: '﻿<?xml version="1.0" encoding="utf-8"?>
```
So I had to make a workaround

File: src/main/java/com/artpie/nuget/PackageIdentity.java (hunk 4 above)
Comment (olegmoz) [reply src/main/java/com/artpie/nuget/PackageIdentity.java]: Read this article, strongly disagree with it though. I think that local variables make code easier to read.

**Comment by olegmoz** [comment]
@yegor256 please review the fix. I have removed local variables and simplified resource reading via cactoos.

File: src/main/java/com/artpie/nuget/Nuspec.java (hunk 3 above)
Comment (yegor256) [reply src/main/java/com/artpie/nuget/Nuspec.java]: @olegmoz maybe something is wrong with the content? Internally it does exactly what you are doing here.

File: src/main/java/com/artpie/nuget/Nuspec.java [hunk 6]
```diff
@@ -48,6 +74,52 @@ public PackageIdentity identity() {
      * @param storage Storage to use for saving.
      */
     public void save(final BlockingStorage storage) {
-        throw new UnsupportedOperationException("Not implemented");
+        final byte[] bytes;
+        try {
+            bytes = this.content.read();
+        } catch (final IOException ex) {
```
Comment (yegor256) [inline src/main/java/com/artpie/nuget/Nuspec.java]: @olegmoz let it fly to the higher level via `throws` declaration. The code will be shorter here.

File: src/test/java/com/artpie/nuget/NewtonJsonResource.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        try {
+            return ByteStreams.toByteArray(
+                new ResourceOf(String.format("newtonsoft.json/12.0.3/%s", this.name)).stream()
+            );
+        } catch (final IOException ex) {
```
Comment (yegor256) [inline src/test/java/com/artpie/nuget/NewtonJsonResource.java]: @olegmoz let it fly to the higher level via `throws` declaration. The code will be shorter here.

**Comment by yegor256** [comment]
@olegmoz see above

**Comment by olegmoz** [comment]
@yegor256 added IOException propagation, please review

**Comment by 0crat** [comment]
@carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by carlosmiranda** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/artipie/nuget-adapter/issues/10#issuecomment-583928682))

@carlosmiranda This is what I know about this job in [CT2E6TK9B](https://www.0crat.com/p/CT2E6TK9B), as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#10](https://github.com/artipie/nuget-adapter/issues/10) is [in scope](http://www.zerocracy.com/policy.html#14) for 6days
  * The role is `REV`
  * The job is assigned to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda) for [5days](http://www.zerocracy.com/policy.html#8)
  * There is a monetary reward attached
  * @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda) will get 25% of their hourly rate for this job
  * The job doesn't have any [impediments](http://www.zerocracy.com/policy.html#9)
  * The [budget](http://www.zerocracy.com/policy.html#4) is 15 minutes/points
  * These users are banned and won't be assigned:
    * @olegmoz[/z](https://www.0crat.com/u/olegmoz): This user reported the ticket
    * @iakunin[/z](https://www.0crat.com/u/iakunin): User was resigned from the ticket
  * Job [footprint](https://www.0crat.com/footprint/CT2E6TK9B?q=%7Bjob%3A%27gh%3Aartipie%2Fnuget-adapter%2310%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by carlosmiranda** [comment]
@yegor256 @g4s8 can we move this ticket out of scope? Seems like Zerocracy missed this being merged.

**Comment by g4s8** [comment]
@0crat out

**Comment by 0crat** [comment]
Order was finished: +15 point(s) just awarded to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda)

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Pull request merged**